packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +50,430 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
-    assert(item.type === "object" || !item.objectType, `Unexpected nested object type ${item.objectType}`);
-    return {
-      type: "list",
-      objectType: item.type === "object" ? item.objectType : item.type,
-      optional: item.type === "object" ? false : !!item.optional,
-    };
-  } else if (type.endsWith("<>")) {
-    assert(allowObjectType, "Expected no 'objectType' in property schema, when using '<>' shorthand");
-    const itemType = type.substring(0, type.length - 2);
-    // Item type defaults to mixed
-    const item: ObjectSchemaProperty = itemType ? normalizePropertyType(itemType) : { type: "mixed" };
-    return {
-      type: "set",
-      objectType: item.type === "object" ? item.objectType : item.type,
-      optional: item.type === "object" ? true : !!item.optional,
-    };
-  } else if (type.endsWith("{}")) {
-    assert(allowObjectType, "Expected no 'objectType' in property schema, when using '{}' shorthand");
-    const itemType = type.substring(0, type.length - 2);
-    // Item type defaults to mixed
-    const item: ObjectSchemaProperty = itemType ? normalizePropertyType(itemType) : { type: "mixed", optional: true };
-    return {
-      type: "dictionary",
-      objectType: item.type === "object" ? item.objectType : item.type,
-      optional: item.type === "object" ? true : !!item.optional,
-    };
-  } else if (type.endsWith("?")) {
-    return {
-      optional: true,
-      type: type.substring(0, type.length - 1),
-    };
-  } else if (type in TYPE_MAPPINGS) {
-    if (type === "dictionary") {
-      return { type, objectType: "mixed", optional: true };
-    } else if (type === "set") {
-      return { type, objectType: "mixed", optional: true };
-    } else if (type === "mixed") {
-      return {type, optional: true};
-    } else {
-      // This type is directly mappable, so it can't be the name a user defined object schema.
-      return { type };
+      "To define an inverse relationship, use { type: 'linkingObjects', objectType: 'ObjectName', property: 'ObjectProperty' }",
+    );
+  } else {
+    // User-defined types
+    objectType = schema;
+    if (!isCollection(type)) {
+      type = "object";
     }
+  }
+
+  if (optionalIsImplicitlyTrue(type, objectType)) {
+    optional = true;
+  } else if (optionalIsImplicitlyFalse(type, objectType)) {
+    assert(
+      !optional,
+      errMessage(
+        name,
+        "Being optional is always 'false' for user-defined types in lists and sets and cannot be set to 'true'. Remove '?' or change the type.",
+      ),
+    );
+    optional = false;
+  }
+
+  const normalizedSchema: CanonicalObjectSchemaProperty = {
+    name: name.propertyName,
+    type: type as PropertyTypeName,
+    optional,
+    indexed: false,
+    mapTo: name.propertyName,
+  };
+  // Add optional properties only if defined (tests expect no 'undefined' properties)
+  objectType !== undefined && (normalizedSchema.objectType = objectType);
```

Comment:
Would love for ECMAScript to add some neat way of doing this directly when the properties are set. Some short syntax for *if the value is undefined, don't add the key*.

Reply:
Really, we shouldn't care about present-but-undefined fields. In basically all cases they should be handled identically to not-present fields.

Internally, JS engines really hate dealing with objects with different shapes which often triggers a deoptimization, so I doubt we will see them making it _easier_ to get there. Not that it matters in this case.

Reply:
Interesting point about the deoptimization.

I don't think it actually matters if undefined fields are passed to Core right? In that case, we can just adjust the tests to expect the undefined fields.

Reply:
I wouldn't worry about the deoptimization here. This specific code probably won't run enough times to trigger JIT specialization optimizations anyway. I was just mentioning as motivation for why we probably won't see it in the future from JS.

As for core, there is no "undefined" in the C++ api, nor can C++ structs have "missing fields". At the bindgen layer we treat both `undefined` and missing on JS record fields the same (by doing the [equivalent](https://github.com/realm/realm-js/blob/63000c018517e6d42508be5f2ade95c9732bcb2a/packages/bindgen/src/templates/node.ts#L674-L684) of `const val = rec.field; if (val !== undefined) { ... } `) and leave the field with its default value. However, in this case there is an extra post-processing layer between this code and the generated binding to convert from something resembling the JS public API to something resembling the C++ API which uses numeric types + bitfields to represent the built-in types rather than strings.